In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [36]:
delv=pd.read_csv(r"D:\hackthon\deliveries.csv")
matc=pd.read_csv(r"D:\hackthon\matches.csv")
#Reading the files

In [4]:
delv

match_id  inning         batting_team                 bowling_team  \
0              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
1              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
2              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
3              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
4              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
...          ...     ...                  ...                          ...   
179073     11415       2  Chennai Super Kings               Mumbai Indians   
179074     11415       2  Chennai Super Kings               Mumbai Indians   
179075     11415       2  Chennai Super Kings               Mumbai Indians   
179076     11415       2  Chennai Super Kings               Mumbai Indians   
179077     11415       2  Chennai Super Kings               Mumbai Indians   

        over  ball    batsman non_striker      bowler  is_super_over  ...  \
0          1     1  DA Warner    S Dhawan    TS Mills              0  ...   
1          1     2  DA Warner    S Dhawan    TS Mills              0  ...   
2          1     3  DA Warner    S Dhawan    TS Mills              0  ...   
3          1     4  DA Warner    S Dhawan    TS Mills              0  ...   
4          1     5  DA Warner    S Dhawan    TS Mills              0  ...   
...      ...   ...        ...         ...         ...            ...  ...   
179073    20     2  RA Jadeja   SR Watson  SL Malinga              0  ...   
179074    20     3  SR Watson   RA Jadeja  SL Malinga              0  ...   
179075    20     4  SR Watson   RA Jadeja  SL Malinga              0  ...   
179076    20     5  SN Thakur   RA Jadeja  SL Malinga              0  ...   
179077    20     6  SN Thakur   RA Jadeja  SL Malinga              0  ...   

        bye_runs  legbye_runs  noball_runs  penalty_runs  batsman_runs  \
0              0            0            0             0             0   
1              0            0            0             0             0   
2              0            0            0             0             4   
3              0            0            0             0             0   
4              0            0            0             0             0   
...          ...          ...          ...           ...           ...   
179073         0            0            0             0             1   
179074         0            0            0             0             2   
179075         0            0            0             0             1   
179076         0            0            0             0             2   
179077         0            0            0             0             0   

        extra_runs  total_runs  player_dismissed dismissal_kind    fielder  
0                0           0               NaN            NaN        NaN  
1                0           0               NaN            NaN        NaN  
2                0           4               NaN            NaN        NaN  
3                0           0               NaN            NaN        NaN  
4                2           2               NaN            NaN        NaN  
...            ...         ...               ...            ...        ...  
179073           0           1               NaN            NaN        NaN  
179074           0           2               NaN            NaN        NaN  
179075           0           1         SR Watson        run out  KH Pandya  
179076           0           2               NaN            NaN        NaN  
179077           0           0         SN Thakur            lbw        NaN  

[179078 rows x 21 columns]

In [5]:
matc

id  season           city        date                        team1  \
0        1    2017      Hyderabad  2017-04-05          Sunrisers Hyderabad   
1        2    2017           Pune  2017-04-06               Mumbai Indians   
2        3    2017         Rajkot  2017-04-07                Gujarat Lions   
3        4    2017         Indore  2017-04-08       Rising Pune Supergiant   
4        5    2017      Bangalore  2017-04-08  Royal Challengers Bangalore   
..     ...     ...            ...         ...                          ...   
751  11347    2019         Mumbai    05/05/19        Kolkata Knight Riders   
752  11412    2019        Chennai    07/05/19          Chennai Super Kings   
753  11413    2019  Visakhapatnam    08/05/19          Sunrisers Hyderabad   
754  11414    2019  Visakhapatnam    10/05/19               Delhi Capitals   
755  11415    2019      Hyderabad    12/05/19               Mumbai Indians   

                           team2                  toss_winner toss_decision  \
0    Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1         Rising Pune Supergiant       Rising Pune Supergiant         field   
2          Kolkata Knight Riders        Kolkata Knight Riders         field   
3                Kings XI Punjab              Kings XI Punjab         field   
4               Delhi Daredevils  Royal Challengers Bangalore           bat   
..                           ...                          ...           ...   
751               Mumbai Indians               Mumbai Indians         field   
752               Mumbai Indians          Chennai Super Kings           bat   
753               Delhi Capitals               Delhi Capitals         field   
754          Chennai Super Kings          Chennai Super Kings         field   
755          Chennai Super Kings               Mumbai Indians           bat   

     result  dl_applied                       winner  win_by_runs  \
0    normal           0          Sunrisers Hyderabad           35   
1    normal           0       Rising Pune Supergiant            0   
2    normal           0        Kolkata Knight Riders            0   
3    normal           0              Kings XI Punjab            0   
4    normal           0  Royal Challengers Bangalore           15   
..      ...         ...                          ...          ...   
751  normal           0               Mumbai Indians            0   
752  normal           0               Mumbai Indians            0   
753  normal           0               Delhi Capitals            0   
754  normal           0          Chennai Super Kings            0   
755  normal           0               Mumbai Indians            1   

     win_by_wickets player_of_match  \
0                 0    Yuvraj Singh   
1                 7       SPD Smith   
2                10         CA Lynn   
3                 6      GJ Maxwell   
4                 0       KM Jadhav   
..              ...             ...   
751               9       HH Pandya   
752               6        AS Yadav   
753               2         RR Pant   
754               6    F du Plessis   
755               0       JJ Bumrah   

                                         venue         umpire1  \
0    Rajiv Gandhi International Stadium, Uppal     AY Dandekar   
1      Maharashtra Cricket Association Stadium  A Nand Kishore   
2       Saurashtra Cricket Association Stadium     Nitin Menon   
3                       Holkar Cricket Stadium    AK Chaudhary   
4                        M Chinnaswamy Stadium             NaN   
..                                         ...             ...   
751                           Wankhede Stadium   Nanda Kishore   
752                  M. A. Chidambaram Stadium     Nigel Llong   
753                           ACA-VDCA Stadium             NaN   
754                           ACA-VDCA Stadium   Sundaram Ravi   
755         Rajiv Gandhi Intl. Cricket Stadium     Nitin Menon   

            umpire2            

In [121]:
matc["dl_applied"].value_counts()


0    737
1     19
Name: dl_applied, dtype: int64

1. checking for question 1 what dl_applied is for

In [9]:
delv.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder'],
      dtype='object')

In [12]:
matc.isnull().sum()

id                   0
season               0
city                 7
date                 0
team1                0
team2                0
toss_winner          0
toss_decision        0
result               0
dl_applied           0
winner               4
win_by_runs          0
win_by_wickets       0
player_of_match      4
venue                0
umpire1              2
umpire2              2
umpire3            637
dtype: int64

2. most null values is seen in city column

In [17]:
matc["city"].value_counts()

Mumbai            101
Kolkata            77
Delhi              74
Bangalore          66
Hyderabad          64
Chennai            57
Jaipur             47
Chandigarh         46
Pune               38
Durban             15
Bengaluru          14
Visakhapatnam      13
Centurion          12
Ahmedabad          12
Rajkot             10
Mohali             10
Indore              9
Dharamsala          9
Johannesburg        8
Cuttack             7
Ranchi              7
Port Elizabeth      7
Cape Town           7
Abu Dhabi           7
Sharjah             6
Raipur              6
Kochi               5
Kanpur              4
Nagpur              3
Kimberley           3
East London         3
Bloemfontein        2
Name: city, dtype: int64

3. Mumbai hosted more number of matches which is 101

In [18]:
matc["season"].value_counts()

2013    76
2012    74
2011    73
2010    60
2014    60
2016    60
2018    60
2019    60
2017    59
2015    59
2008    58
2009    57
Name: season, dtype: int64

4. most number of matches played is 76 in the year 2013

In [27]:
len(matc.loc[(matc["season"]==2011 )&(matc["winner"]=="Mumbai Indians")])

10

5. Maximum wins by Mumbai Indians in 2011 are 10


In [29]:
matc["toss_decision"].value_counts()

field    463
bat      293
Name: toss_decision, dtype: int64

6. checking what kind of data for 6th question

In [34]:
463/len(matc["toss_decision"])*100


61.24338624338625

In [35]:
293/len(matc["toss_decision"])*100


38.75661375661375

7. 38.7 & 61.2 is the percentage of getting bat and field in feature 'toss_descision'

In [54]:
field=0
bat=0
for x,y in matc.iterrows():
    if y["toss_decision"]=="field":
        if y["toss_winner"]== y["winner"]:
            field+=1
    else:
        if y["toss_winner"]== y["winner"]:
            
            bat+=1

In [55]:
print("Total wins when fielded", field)
print("Total wins when bated", bat)

Total wins when fielded 259
Total wins when bated 134


8. Teams winning tosses and electing field first have won more number of matches.

In [73]:
#In which years where taking batting second have won more number of matches?
year=[]
for x,y in matc.iterrows():
    if y["toss_decision"]=="field":
        if y["team1"]==y["winner"] or y["team2"]==y["winner"]:
            year.append(y["season"])
        



In [74]:
df=pd.Series(year)

In [75]:
df.value_counts()

2018    50
2016    49
2019    49
2017    48
2011    48
2014    41
2012    37
2008    32
2015    32
2013    31
2009    22
2010    21
dtype: int64

9. 2018 and 2016 are the years where taking batting second have won more number of matches

In [77]:
mdf = pd.merge(delv, matc, left_on='match_id', right_on='id')

In [78]:
mdf

match_id  inning         batting_team                 bowling_team  \
0              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
1              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
2              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
3              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
4              1       1  Sunrisers Hyderabad  Royal Challengers Bangalore   
...          ...     ...                  ...                          ...   
179073     11415       2  Chennai Super Kings               Mumbai Indians   
179074     11415       2  Chennai Super Kings               Mumbai Indians   
179075     11415       2  Chennai Super Kings               Mumbai Indians   
179076     11415       2  Chennai Super Kings               Mumbai Indians   
179077     11415       2  Chennai Super Kings               Mumbai Indians   

        over  ball    batsman non_striker      bowler  is_super_over  ...  \
0          1     1  DA Warner    S Dhawan    TS Mills              0  ...   
1          1     2  DA Warner    S Dhawan    TS Mills              0  ...   
2          1     3  DA Warner    S Dhawan    TS Mills              0  ...   
3          1     4  DA Warner    S Dhawan    TS Mills              0  ...   
4          1     5  DA Warner    S Dhawan    TS Mills              0  ...   
...      ...   ...        ...         ...         ...            ...  ...   
179073    20     2  RA Jadeja   SR Watson  SL Malinga              0  ...   
179074    20     3  SR Watson   RA Jadeja  SL Malinga              0  ...   
179075    20     4  SR Watson   RA Jadeja  SL Malinga              0  ...   
179076    20     5  SN Thakur   RA Jadeja  SL Malinga              0  ...   
179077    20     6  SN Thakur   RA Jadeja  SL Malinga              0  ...   

        result  dl_applied               winner  win_by_runs  win_by_wickets  \
0       normal           0  Sunrisers Hyderabad           35               0   
1       normal           0  Sunrisers Hyderabad           35               0   
2       normal           0  Sunrisers Hyderabad           35               0   
3       normal           0  Sunrisers Hyderabad           35               0   
4       normal           0  Sunrisers Hyderabad           35               0   
...        ...         ...                  ...          ...             ...   
179073  normal           0       Mumbai Indians            1               0   
179074  normal           0       Mumbai Indians            1               0   
179075  normal           0       Mumbai Indians            1               0   
179076  normal           0       Mumbai Indians            1               0   
179077  normal           0       Mumbai Indians            1               0   

        player_of_match                                      venue  \
0          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
2          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
3          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
4          Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
...                 ...                                        ...   
179073        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179074        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179075        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179076        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   
179077        JJ Bumrah         Rajiv Gandhi Intl. Cricket Stadium   

            umpire1    umpire2      umpire3  
0       AY Dandekar   NJ Llong          NaN  
1       AY Dandekar   NJ Llong          NaN  
2       AY Dandekar   NJ Llong          NaN  
3       AY Dandekar   NJ Llong          NaN  
4       AY Dandekar   NJ Llong          NaN  
...             ...        ...          ...  
179073  Nitin Menon  I

In [80]:
mdf.columns

Index(['match_id', 'inning', 'batting_team', 'bowling_team', 'over', 'ball',
       'batsman', 'non_striker', 'bowler', 'is_super_over', 'wide_runs',
       'bye_runs', 'legbye_runs', 'noball_runs', 'penalty_runs',
       'batsman_runs', 'extra_runs', 'total_runs', 'player_dismissed',
       'dismissal_kind', 'fielder', 'id', 'season', 'city', 'date', 'team1',
       'team2', 'toss_winner', 'toss_decision', 'result', 'dl_applied',
       'winner', 'win_by_runs', 'win_by_wickets', 'player_of_match', 'venue',
       'umpire1', 'umpire2', 'umpire3'],
      dtype='object')

In [82]:
#In 2019 which batsman scored highest number of runs by hitting 6's and 4's?
batmen=mdf.loc[(mdf["season"]== 2019) & (mdf["batsman_runs"]==6) |(mdf["batsman_runs"]==4)]

In [84]:
batmen["batsman"].value_counts()

S Dhawan       537
SK Raina       504
V Kohli        495
G Gambhir      492
DA Warner      480
              ... 
GB Hogg          1
A Zampa          1
DM Bravo         1
A Choudhary      1
TA Boult         1
Name: batsman, Length: 413, dtype: int64

11. In 2019 the batsman who scored highest number of runs by hitting 6's and 4's is S Dhawan

In [119]:
#12. Most number of wickets taken by a bowler is


In [110]:
pd.set_option('display.max_columns', None)

In [116]:
wick=delv.loc[delv["player_dismissed"].notna()]

In [118]:
wick["bowler"].value_counts()

SL Malinga         188
DJ Bravo           168
A Mishra           165
Harbhajan Singh    161
PP Chawla          156
                  ... 
CL White             1
BMAJ Mendis          1
LA Carseldine        1
GS Sandhu            1
AM Rahane            1
Name: bowler, Length: 352, dtype: int64

12. Most number of wickets taken by a bowler is SL Malinga

In [123]:
#13. What is the strike rate of Kohli in 2016


In [126]:
delv["batsman"].value_counts()

V Kohli          4211
SK Raina         4044
RG Sharma        3816
S Dhawan         3776
G Gambhir        3524
                 ... 
K Ahmed             1
U Kaul              1
DP Vijaykumar       1
P Raj               1
LH Ferguson         1
Name: batsman, Length: 516, dtype: int64

In [155]:
kohli=mdf.loc[(mdf["batsman"]=="V Kohli") & (mdf["season"]==2016)]

In [160]:
kohli["batsman_runs"].sum()/len(kohli)*100

148.54961832061068

In [157]:
kohli

match_id  inning                 batting_team         bowling_team  \
137017       580       1  Royal Challengers Bangalore  Sunrisers Hyderabad   
137018       580       1  Royal Challengers Bangalore  Sunrisers Hyderabad   
137019       580       1  Royal Challengers Bangalore  Sunrisers Hyderabad   
137020       580       1  Royal Challengers Bangalore  Sunrisers Hyderabad   
137026       580       1  Royal Challengers Bangalore  Sunrisers Hyderabad   
...          ...     ...                          ...                  ...   
150410       636       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
150413       636       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
150414       636       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
150415       636       2  Royal Challengers Bangalore  Sunrisers Hyderabad   
150416       636       2  Royal Challengers Bangalore  Sunrisers Hyderabad   

        over  ball  batsman     non_striker        bowler  is_super_over  \
137017     1     3  V Kohli        CH Gayle       A Nehra              0   
137018     1     4  V Kohli        CH Gayle       A Nehra              0   
137019     1     5  V Kohli        CH Gayle       A Nehra              0   
137020     1     6  V Kohli        CH Gayle       A Nehra              0   
137026     2     6  V Kohli  AB de Villiers       B Kumar              0   
...      ...   ...      ...             ...           ...            ...   
150410    12     5  V Kohli  AB de Villiers  Bipul Sharma              0   
150413    13     2  V Kohli  AB de Villiers       BB Sran              0   
150414    13     3  V Kohli  AB de Villiers       BB Sran              0   
150415    13     4  V Kohli  AB de Villiers       BB Sran              0   
150416    13     5  V Kohli  AB de Villiers       BB Sran              0   

        wide_runs  bye_runs  legbye_runs  noball_runs  penalty_runs  \
137017          0         0            0            0             0   
137018          0         0            0            0             0   
137019          0         0            0            0             0   
137020          0         0            0            0             0   
137026          0         0            0            0             0   
...           ...       ...          ...          ...           ...   
150410          0         0            0            0             0   
150413          0         0            0            0             0   
150414          0         0            0            0             0   
150415          0         0            0            0             0   
150416          0         0            0            0             0   

        batsman_runs  extra_runs  total_runs player_dismissed dismissal_kind  \
137017             0           0           0              NaN            NaN   
137018             0           0           0              NaN            NaN   
137019             0           0           0              NaN            NaN   
137020             4           0           4              NaN            NaN   
137026             1           0           1              NaN            NaN   
...              ...         ...         ...              ...            ...   
150410             1           0           1              NaN            NaN   
150413             6           0           6              NaN            NaN   
150414             2           0           2              NaN            NaN   
150415             2           0           2              NaN            NaN   
150416             0           0           0          V Kohli         bowled   

       fielder   id  season       city        date  \
137017     NaN  580    2016  Bangalore  2016-04-12   
137018     NaN  580    2016  Bangalore  2016-04-12   
137019     NaN  580    2016  Bangalore  2016-04-12   
137020     NaN  580    2016  Bangalore  2016-04-12   
137026     NaN  580    2016  Bangalore  2016-04-12   
...        ...  ... 

In [167]:
#14.Bowlers with maximum number of extras:


In [163]:
extra=delv.loc[delv["extra_runs"] != 0]

In [166]:
extra["bowler"].value_counts()

SL Malinga         221
P Kumar            185
UT Yadav           180
DJ Bravo           176
B Kumar            162
                  ... 
MJ Henry             1
Swapnil Singh        1
Sunny Gupta          1
M Vijay              1
Gagandeep Singh      1
Name: bowler, Length: 357, dtype: int64

In [179]:
id1=extra.groupby('bowler')['extra_runs'].sum().max()

In [185]:
id1=extra.groupby('bowler')['extra_runs'].sum()

In [186]:
df1=pd.DataFrame({'bowler': id1.index, 'runs': id1.values})

In [190]:
df1.loc[df1["runs"]==df1["runs"].max()]

bowler  runs
302  SL Malinga   293

14. Bowlers with maximum number of extras is malinga

In [192]:
#15. Which venue has hosted most number of IPL matches?


In [199]:
matc["venue"].value_counts()

Eden Gardens                                            77
M Chinnaswamy Stadium                                   73
Wankhede Stadium                                        73
Feroz Shah Kotla                                        67
Rajiv Gandhi International Stadium, Uppal               56
MA Chidambaram Stadium, Chepauk                         49
Sawai Mansingh Stadium                                  47
Punjab Cricket Association Stadium, Mohali              35
Maharashtra Cricket Association Stadium                 21
Subrata Roy Sahara Stadium                              17
Dr DY Patil Sports Academy                              17
Kingsmead                                               15
Punjab Cricket Association IS Bindra Stadium, Mohali    14
SuperSport Park                                         12
Sardar Patel Stadium, Motera                            12
Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium     11
Brabourne Stadium                                       

15. Its Eden Gardens, which hosted more number of matches

In [200]:
#16. In 2017 when sunrisers hyderabad clashed against Royal Challengers Bangalore which team player won player of the match?

In [213]:

filtered_matches = matc.loc[((matc["team1"].isin(["Sunrisers Hyderabad", "Royal Challengers Bangalore"])) & 
                              (matc["team2"].isin(["Sunrisers Hyderabad", "Royal Challengers Bangalore"])) & 
                              (matc["season"] == 2017))]


filtered_matches["player_of_match"].values[0]



'Yuvraj Singh'

16. Its seems yuvaraj singh

In [238]:
#17. Across seasons who are the top three batsman's with most number of run out?

In [234]:
runout=delv.loc[delv["dismissal_kind"]=="run out"]

In [237]:
runout["batsman"].value_counts()

MS Dhoni       21
RG Sharma      17
KD Karthik     16
G Gambhir      15
V Kohli        15
               ..
RR Bhatkal      1
T Kohli         1
PC Valthaty     1
SP Fleming      1
J Suchith       1
Name: batsman, Length: 277, dtype: int64

Across seasons the top three batsman's with most number of run out is 
MS Dhoni       21
RG Sharma      17
KD Karthik     16

In [239]:
#18. What are the total runs scored by V Kohli when the bowler was JJ Burmah?

In [243]:
delv.loc[(delv["batsman"]=="V Kohli") & (delv["bowler"]=="JJ Burmah")]

Empty DataFrame
Columns: [match_id, inning, batting_team, bowling_team, over, ball, batsman, non_striker, bowler, is_super_over, wide_runs, bye_runs, legbye_runs, noball_runs, penalty_runs, batsman_runs, extra_runs, total_runs, player_dismissed, dismissal_kind, fielder]
Index: []

In [248]:
delv["bowler"].value_counts.unique()

AttributeError: 'function' object has no attribute 'unique'